File: go-algorithm/leetcode/problem4.go
package main

// 4. median of two sorted arrays
// https://leetcode.com/problems/median-of-two-sorted-arrays/

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	i, j, m, n := 0, 0, len(nums1), len(nums2)
	merged := make([]int, 0)
	for i < m && j < n {
		if nums1[i] <= nums2[j] {
			merged = append(merged, nums1[i])
			i++
		} else {
			merged = append(merged, nums2[j])
			j++
		}
	}
	if i < m {
		merged = append(merged, nums1[i:]...)
	}
	if j < n {
		merged = append(merged, nums2[j:]...)
	}
	if (m+n)%2 == 1 {
		return float64(merged[(m+n)/2])
	}
	mid0, mid1 := (m+n)/2-1, (m+n)/2
	return float64(merged[mid0]+merged[mid1]) / 2
}

File: python-algo/leetcode/problem_4.py
"""4. Median of Two Sorted Arrays
https://leetcode.com/problems/median-of-two-sorted-arrays/
"""
from typing import List


class Solution:
    def find_median_sorted_arrays(self, nums1: List[int], nums2: List[int]) -> float:
        n, m = len(nums1), len(nums2)
        x, y = divmod(n + m, 2)
        is_odd = True if y == 1 else False
        mid_l = mid_r = x
        if not is_odd:
            mid_l = mid_r - 1
        i = j = 0
        counter = -1
        ans = 0
        while i < n or j < m:
            if i == n and j < m:
                smaller = nums2[j]
                j += 1
            elif i < n and j == m:
                smaller = nums1[i]
                i += 1
            elif nums1[i] <= nums2[j]:
                smaller = nums1[i]
                i += 1
            else:
                smaller = nums2[j]
                j += 1
            counter += 1
            if counter == mid_l:
                if is_odd:
                    return smaller
                else:
                    ans = smaller
            if counter == mid_r:
                return (ans + smaller) / 2
